File: services/forecast-engine/src/forecast_engine/messaging/ingestion.py
"""Schema de payload de ingestion.request — espelha o schema Zod.

Ver: specs/001-sop-cycle-forecasting/contracts/messaging.md
"""

from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, field_validator


class IngestionRequestPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    jobId: UUID  # noqa: N815
    scenarioId: UUID  # noqa: N815
    kind: Literal["SALES_HISTORY"]
    objectUri: str  # noqa: N815
    declaredLabels: list[str]  # noqa: N815
    uploadedById: UUID  # noqa: N815

    @field_validator("objectUri")
    @classmethod
    def _s3_uri(cls, v: str) -> str:
        import re
        if not re.match(r"^s3://[a-z0-9.\-]+/.+$", v):
            raise ValueError(f"esperado um URI s3://bucket/caminho, recebido {v!r}")
        return v

    @field_validator("declaredLabels")
    @classmethod
    def _nao_vazio(cls, v: list[str]) -> list[str]:
        if not v:
            raise ValueError("declaredLabels não pode ser vazio")
        return v

File: packages/contracts/src/messaging/ingestion.ts
import { z } from 'zod';
import { ObjectUri } from './envelope.js';

/**
 * Schema de payload de `ingestion.request`.
 *
 * Ver: specs/001-sop-cycle-forecasting/contracts/messaging.md
 */

const uuid = z.string().uuid();

export const IngestionKind = z.enum(['SALES_HISTORY', 'COLLABORATION_SHEET', 'ACTUAL_SALES']);
export type IngestionKind = z.infer<typeof IngestionKind>;

export const IngestionRequestPayload = z.object({
  jobId: uuid,
  scenarioId: uuid,
  kind: IngestionKind,
  objectUri: ObjectUri,
  declaredLabels: z.array(z.string()).min(1),
  uploadedById: uuid,
});
export type IngestionRequestPayload = z.infer<typeof IngestionRequestPayload>;
